sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py

```diff
@@ -4,61 +4,99 @@
-    the AAD Token, refresh token, and access token before performing a call to service.
-
-    :param credential: Azure Token Credential for authenticating with Azure
-    :type credential: TokenCredential
-    """
-
-    BEARER = "Bearer"
-    AUTHENTICATION_CHALLENGE_PARAMS_PATTERN = re.compile('(?:(\\w+)="([^""]*)")+')
-    WWW_AUTHENTICATE = "WWW-Authenticate"
-    SCOPE_PARAMETER = "scope"
-    SERVICE_PARAMETER = "service"
-    AUTHORIZATION = "Authorization"
-
-    def __init__(self, credential, url, pipeline):
-        self._credential = credential
-        self.url = url
-        self.pipeline = pipeline
+        """Called before the policy sends a request.
+        The base implementation authorizes the request with a bearer token.
+        :param ~azure.core.pipeline.PipelineRequest request: the request
+        """
+
+        if self._token is None or self._need_new_token():
+            self._token = self._credential.get_token(self._scopes)
+            try:
+                self._token = self._token.token
+            except AttributeError:
+                pass
+        request.http_request.headers["Authorization"] = "Bearer " + self._token
+
+    def send(self, request):
+        # type: (PipelineRequest) -> PipelineResponse
+        """Authorizes a request with a bearer token, possibly handling an authentication challenge
+        :param ~azure.core.pipeline.PipelineRequest request: the request
+        """
+        _enforce_https(request)
+
+        self.on_request(request)
+
+        response = self.next.send(request)
+
+        if response.http_response.status_code == 401:
+            self._token = None  # any cached token is invalid
+            challenge = response.http_response.headers.get("WWW-Authenticate")
+            if challenge and self.on_challenge(request, response, challenge):
+                response = self.next.send(request)
+
+        return response
+
+    def on_challenge(self, request, response, challenge):
+        # type: (PipelineRequest, PipelineResponse, str) -> bool
+        """Authorize request according to an authentication challenge
+        This method is called when the resource provider responds 401 with a WWW-Authenticate header.
+        :param ~azure.core.pipeline.PipelineRequest request: the request which elicited an authentication challenge
+        :param ~azure.core.pipeline.PipelineResponse response: the resource provider's response
+        :param str challenge: response's WWW-Authenticate header, unparsed. It may contain multiple challenges.
+        :returns: a bool indicating whether the policy should send the request
+        """
+        # pylint:disable=unused-argument,no-self-use
+
+        access_token = self._exchange_client.get_acr_access_token(challenge)
+        self._token = access_token
+        request.http_request.headers["Authorization"] = "Bearer " + self._token
+        return access_token is not None
```

Comment:
Is it possible for `get_acr_access_token` to return None? The line above raises when it does.